### PR TITLE
[RLlib] Introduces Discrete/Gaussian/Deterministic action spaces to Catalog + tests

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

In order to [flip the switch](https://github.com/ray-project/ray/pull/32808) for RLModules, we need to support a series of action distributions. This PR introduces these action distributions to the Catalog, tests and introduces them to the PPORLModule.

### CODE REVIEW

File: rllib/core/models/catalog.py [hunk 1]
```diff
@@ -224,3 +240,80 @@ def get_tokenizer_config(
                 **{"use_lstm": False, "use_attention": False},
             },
         )
+
+    @classmethod
+    def get_action_dist_cls_dict(
```
Comment (kouroshHakha) [inline rllib/core/models/catalog.py:245]: Should this just take framework as input? and return the corresponding the  dist class? instead of a mapping from fw to the class?

File: rllib/core/models/catalog.py [hunk 2]
```diff
@@ -224,3 +240,80 @@ def get_tokenizer_config(
...
+        You can get the required input dimension for the distribution by calling
+        `action_dict_cls.required_model_output_shape(action_space, model_config_dict)`
+        on the retrieved class. This is useful, because the Catalog needs to find out
+        about the required input dimension for the distribution before the model that
+        outputs these inputs is configured.
+
+        Args:
+            action_space: Action space of the target gym env.
+            model_config_dict: Optional model config.
+            deterministic: Whether to return a Deterministic distribution instead of
```
Comment (kouroshHakha) [inline rllib/core/models/catalog.py:262]: better wording: "Whether to return a Deterministic distribution on input logits instead of a stochastic distributions. For example for Discrete spaces, the stochastic is a Categorical distribution with output logits, while the deterministic distribution will be to output the argmax of logits directly."

File: rllib/core/models/catalog.py [hunk 3]
```diff
@@ -224,3 +240,80 @@ def get_tokenizer_config(
...
+        """
+        assert not model_config_dict.get("custom_action_dist"), (
+            "Specify a custom action "
+            "distribution by overwriting "
+            "Catalog.get_action_distribution."
+        )
+
+        # Box space -> DiagGaussian OR Deterministic.
+        if isinstance(action_space, Box):
+            if action_space.dtype.name.startswith("int"):
```
Comment (kouroshHakha) [inline rllib/core/models/catalog.py:276]: why can't we just check `isinstance(action_space, MultiDiscrete)`?

Comment (kouroshHakha) [reply rllib/core/models/catalog.py:276]: Sorry, I understood the condition now. Isn't there any better api to check the dtype? I particularly don't feel safe about this `.name.startswith("int"):`

File: rllib/core/models/catalog.py [hunk 4]
```diff
@@ -224,3 +240,80 @@ def get_tokenizer_config(
...
+                if len(action_space.shape) > 1:
+                    raise UnsupportedSpaceException(
+                        "Action space has multiple dimensions "
+                        "{}. ".format(action_space.shape)
+                        + "Consider reshaping this into a single dimension, "
+                        "using a custom action distribution, "
+                        "using a Tuple action space, or the multi-agent API."
+                    )
+                if deterministic:
+                    return {"torch": TorchDeterministic, "tf": TfDeterministic}
```
Comment (kouroshHakha) [inline rllib/core/models/catalog.py:291]: I would predefine all these dicts at the top of this function and then index via framework input arg.

File: rllib/core/models/catalog.py [hunk 5]
```diff
@@ -224,3 +240,80 @@ def get_tokenizer_config(
                 **{"use_lstm": False, "use_attention": False},
             },
         )
+
+    @classmethod
+    def get_action_dist_cls_dict(
+        cls,
+        action_space: gym.Space,
+        model_config_dict: ModelConfigDict,
```
Comment (kouroshHakha) [inline rllib/core/models/catalog.py:248]: I don't see how you use `model_config_dict` here. Is this currently used for inferring the dist classes? I think for customization it's better to have the user override this function of the catalog and return their custom dist class with a custom catalog instead of passing it `model_config_dict`

File: rllib/core/models/tests/test_catalog.py [hunk 6]
```diff
@@ -173,6 +185,76 @@ def test_get_encoder_config(self):
...
+        """Tests if we can create a bunch of action distributions.
+
+        Action distributions are created from the base catalog class. Things this
+        test checks:
+            - Whether we output the correct action distributions classes.
+            - Whether we can instantiate the action distributions, query their
+                required input dimensions and sample from them.
+
+        """
+        action_spaces_dist_types_and_expected_cls_dicts = [
```
Comment (kouroshHakha) [inline rllib/core/models/tests/test_catalog.py:198]: Can you make these a named tuple? for example the boolean in index 1 is not clear :)

Comment (kouroshHakha) [reply rllib/core/models/tests/test_catalog.py:198]: what about discrete + deterministic.

File: rllib/core/models/tests/test_catalog.py [hunk 7]
```diff
@@ -173,6 +185,76 @@ def test_get_encoder_config(self):
...
+                action_space=action_space,
+                model_config_dict=MODEL_DEFAULTS.copy(),
+            )
+            dist_dict = catalog.get_action_dist_cls_dict(
+                action_space=action_space,
+                model_config_dict=MODEL_DEFAULTS.copy(),
+                deterministic=deterministic,
+            )
+
+            for framework, sess in framework_iterator(
```
Comment (kouroshHakha) [inline rllib/core/models/tests/test_catalog.py:232]: why is sess, true?

File: rllib/core/models/tests/test_catalog.py [hunk 8]
```diff
@@ -173,6 +185,76 @@ def test_get_encoder_config(self):
...
+                frameworks=["tf", "torch"], session=True
+            ):
+                if framework not in dist_dict:
```
Comment (kouroshHakha) [inline rllib/core/models/tests/test_catalog.py:235]: Should we skip? or should we raise an erorr?

File: rllib/core/models/tests/test_catalog.py [hunk 9]
```diff
@@ -173,6 +185,76 @@ def test_get_encoder_config(self):
...
+                else:
+                    logits = tf.convert_to_tensor(logits)
+                # We don't need a model if we input tensors
+                dist = dist_cls.from_logits(logits=logits)
+                assert isinstance(dist, expected_cls_dict[framework])
+                outputs = dist.sample()
+                if framework == "torch":
+                    action = outputs.numpy()[0]
+                else:
+                    action = sess.run(outputs)[0]
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/core/models/tests/test_catalog.py:254]: turn off eager tracing and tf2 will have the numpy values.

[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/core/models/tests/test_catalog.py:254]: if that's what you are trying to fix here.

File: rllib/algorithms/ppo/torch/ppo_torch_rl_module.py [hunk 10]
```diff
@@ -40,6 +40,11 @@ def __init__(self, *args, **kwargs):
         TorchRLModule.__init__(self, *args, **kwargs)
         PPORLModuleBase.__init__(self, *args, **kwargs)
 
+        if self._is_discrete:
```
Comment (kouroshHakha) [inline rllib/algorithms/ppo/torch/ppo_torch_rl_module.py:43]: We need to remove this stuff, once the other PR gets merged first.

File: rllib/algorithms/ppo/ppo_base_rl_module.py [hunk 11]
```diff
@@ -31,6 +31,8 @@ def __init__(self, config: RLModuleConfig):
         self.pi = catalog.build_pi_head(framework=self.framework)
         self.vf = catalog.build_vf_head(framework=self.framework)
 
+        self.action_dist_cls = catalog.get_action_dist_cls(framework=self.framework)
```
Comment (kouroshHakha) [inline rllib/algorithms/ppo/ppo_base_rl_module.py:34]: `get_action_dist_cls` or `get_action_dist_dict`?

Comment (kouroshHakha) [reply rllib/algorithms/ppo/ppo_base_rl_module.py:34]: I think this should become the catalog base API. not a PPO specific API

File: rllib/algorithms/ppo/ppo_catalog.py [hunk 12]
```diff
@@ -79,7 +80,8 @@ def __init__(
             hidden_layer_dims=post_fcnet_hiddens,
             hidden_layer_activation=post_fcnet_activation,
             output_activation="linear",
-            output_dim=pi_output_dim,
+            output_dim=None,  # We don't know the output dimension yet, because it
```
Comment (kouroshHakha) [inline rllib/algorithms/ppo/ppo_catalog.py:83]: can't you just say `output_dim=self.action_dist_cls.required_model_output_shape(
                space=self.action_space, model_config=self.model_config_dict` here?

**Review by kouroshHakha — COMMENTED** [review]
We need to merge the other PR first, for me to see the diffs better. But one major concern popped up here. Shouldn't `get_action_dist_cls(self, framework)` be defined in the base class of Catalog? instead of being a PPO-specific one in this PR?

File: rllib/core/models/catalog.py (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/catalog.py:262]: Thanks putting in the effort of correcting these in my PRs! 🙂

File: rllib/core/models/catalog.py (hunk 3 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/catalog.py:276]: Yeah it's wonky, you are right. We had the same logic in the old catalog. I guess we should go with the "let's not try to be clever" logic here and simply raise an error that tells folks to use a fitting distribution.

Comment (ArturNiederfahrenhorst) [reply rllib/core/models/catalog.py:276]: I don't think there is a standard API, but `action_space.dtype.char in np.typecodes["AllInteger"]` seems like a stable and precise solution.

File: rllib/core/models/catalog.py (hunk 5 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/catalog.py:248]: Yeah absolutely.

Comment (ArturNiederfahrenhorst) [reply rllib/core/models/catalog.py:248]: The only issue is that it is needed for the `required_model_output_shape` methods that we need to compute input shapes for multi-distributions (that are not here yet).

File: rllib/core/models/tests/test_catalog.py (hunk 6 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/tests/test_catalog.py:198]: What do you mean? They are there, no?

File: rllib/algorithms/ppo/ppo_catalog.py (hunk 12 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/ppo/ppo_catalog.py:83]: We don't know the framework yet. So we'd have to instantiate either of the framework-specific classes here, which I guess is a no-no.

File: rllib/algorithms/ppo/torch/ppo_torch_rl_module.py (hunk 10 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/ppo/torch/ppo_torch_rl_module.py:43]: Yep.

File: rllib/core/models/catalog.py (hunk 5 above)
Comment (kouroshHakha) [reply rllib/core/models/catalog.py]: `get_dist_class` won't need model_config but `get_dist_object` will, right?

File: rllib/core/models/tests/test_catalog.py (hunk 6 above)
Comment (kouroshHakha) [reply rllib/core/models/tests/test_catalog.py]: we have Box + True, Box + False, and Disc + None. I think we should have Disc + True and Disc + False. No?

File: rllib/core/models/tests/test_catalog.py (hunk 7 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/tests/test_catalog.py]: My mistake 👍

File: rllib/algorithms/ppo/ppo_catalog.py (hunk 12 above)
Comment (kouroshHakha) [reply rllib/algorithms/ppo/ppo_catalog.py:77]: what is framework specific? `self.action_dist_cls` is just the class.

File: rllib/core/models/tests/test_catalog.py [hunk 13]
```diff
@@ -225,15 +231,14 @@ def test_get_action_dist_cls_dict(self):
...
             dist_dict = catalog.get_action_dist_cls_dict(
                 action_space=action_space,
-                model_config_dict=MODEL_DEFAULTS.copy(),
                 deterministic=deterministic,
             )
 
-            for framework, sess in framework_iterator(
-                frameworks=["tf", "torch"], session=True
+            for framework in framework_iterator(
+                frameworks=["tf2", "torch"], session=False
```
Comment (kouroshHakha) [inline rllib/core/models/tests/test_catalog.py:238]: nit: session can be left as None.

**Review by kouroshHakha — COMMENTED** [review]
Still waiting for the diffs, but I think we are almost good. Just another follow up q.

File: rllib/core/models/tests/test_catalog.py (hunk 6 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/tests/test_catalog.py]: @kouroshHakha Do you want also to return a deterministic distribution in these cases? Because so far, we've always just used "_max(q_values)_" for that sort of stuff.

**Review by gjoliver — APPROVED** [review]
